apps/letterboxd/letterboxd.star

```diff
@@ -0,0 +1,361 @@
+REWATCH_ICON = base64.decode(REWATCH_IMAGE)
+
+PERC_IN = 0.10
+PERC_OUT = 0.90
+DURATION = 105
+DELAY_1 = 0
+DELAY_2 = 95
+DELAY_3 = 190
+ACTIVITY_START_INDEX = 0
+ACTIVITY_END_INDEX = 4
+REVIEW_COUNT = 3
+
+def main(config):
+    username = config.str("username", DEFAULT_USERNAME)
+    lid = getLID(username)
+    activityUrl = "/member/%s/activity" % (lid)
+    memberActivity = getMemberActivity(activityUrl)
+    recentActivity = memberActivity["items"]
+    entryList = recentActivity[ACTIVITY_START_INDEX:ACTIVITY_END_INDEX]
+    entryList = [getEntry(entryList[i]) for i in range(REVIEW_COUNT)]
+    return render.Root(
+        child = renderReviews(entryList),
+    )
+
+# functions
+def get(route, params = ""):
+    method = "GET"
+    toSalt = constructUrl(route, api_key, params)
+    signature = getSignature(method, toSalt)
+
+    url = "%s&signature=%s" % (toSalt, signature)
+    res = http.get(url = url, headers = {"Content-Type": "application/json"})
+
+    if res.status_code != 200:
+        fail("station request failed with status code: %d - %s" %
+             (res.status_code, res.body()))
+
+    jsonresults = res.json()
+    return jsonresults
+
+def getSignature(method, toSalt):
+    body = ""
+    str = "%s\u0000%s\u0000%s" % (method, toSalt, body)
+    sigHash = hmac.sha256(api_secret, str).lower()
+    lowerSig = sigHash.lower()
+    return lowerSig
+
+def constructUrl(authRoute, apiKey = api_key, params = ""):
+    timestamp = time.now().unix
+    nonce = http.get(UUID_URL).json()[0]
+    formattedParams = ["&%s=%s" % (p, params[p]) for p in params] if params else [""]
+    constructParams = "".join(formattedParams)
+    constructed = "%s%s?apikey=%s&nonce=%s&timestamp=%s%s" % (BASE_URL, authRoute, apiKey, nonce, timestamp, constructParams)
+    return constructed
+
+def getLID(target):
+    url = "%s%s" % (LB_URL, target)
+    res = http.get(url = url)
+    headers = res.headers
+    lid = headers["X-Letterboxd-Identifier"]
+    return lid
+
+def getMemberActivity(query):
+    activity = get(query, {"where": "NotOwnActivity", "include": "DiaryEntryActivity"})
+    return activity
+
+def getEntry(item):
+    film = item["diaryEntry"]["film"]
+    diaryEntry = item["diaryEntry"]
+    id = diaryEntry["id"]
+    logEntryUrl = ("/log-entry/%s") % (id)
+    logEntry = get(logEntryUrl)
+
+    review = logEntry.get("review", "")
+    reviewText = review.get("lbml", "") if review != "" else ""
+    movieName = film.get("name", "")
+    releaseYear = film.get("releaseYear", "")
+    releaseYear = int(releaseYear)
+    releaseYear = "%d" % releaseYear
+    diaryDetails = logEntry["diaryDetails"]
+    rewatch = diaryDetails.get("rewatch", False)
+    posterUrl = film["poster"]["sizes"][0]["url"]
+    poster = http.get(posterUrl).body()
+    owner = item["diaryEntry"]["owner"]
+    ownerName = owner["displayName"]
+    avatarUrl = owner["avatar"]["sizes"][0]["url"]
+    avatar = http.get(avatarUrl).body()
```

Comment:
It'd be good to add caching to this and all other HTTP calls. There'll be an awful lot of unnecessary requests going out otherwise.

Reply:
@matslina I think I have all the requests covered with caching now.